backend/hasher/object.go

```diff
@@ -0,0 +1,234 @@
+		return err
+	}
+	defer func() {
+		_ = r.Close()
+	}()
+	if _, err = io.Copy(ioutil.Discard, r); err != nil {
+		fs.Infof(o, "update failed (copy): %v", err)
+		return err
+	}
+	return nil
+}
+
+// Update the object with the given data, time and size.
+func (o *Object) Update(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) error {
+	_ = o.f.pruneHash(src.Remote())
+	return o.Object.Update(ctx, in, src, options...)
+}
+
+// Remove an object.
+func (o *Object) Remove(ctx context.Context) error {
+	_ = o.f.pruneHash(o.Remote())
+	return o.Object.Remove(ctx)
+}
+
+// SetModTime sets the modification time of the file.
+// Also prunes the cache entry when modtime changes so that
+// touching a file will trigger checksum recalculation even
+// on backends that don't provide modTime with fingerprint.
+func (o *Object) SetModTime(ctx context.Context, mtime time.Time) error {
+	if mtime != o.Object.ModTime(ctx) {
+		_ = o.f.pruneHash(o.Remote())
+	}
+	return o.Object.SetModTime(ctx, mtime)
+}
+
+// Open opens the file for read.
+// Full reads will also update object hashes.
+func (o *Object) Open(ctx context.Context, options ...fs.OpenOption) (r io.ReadCloser, err error) {
+	size := o.Size()
+	var offset, limit int64 = 0, -1
+	for _, option := range options {
+		switch opt := option.(type) {
+		case *fs.SeekOption:
+			offset = opt.Offset
+		case *fs.RangeOption:
+			offset, limit = opt.Decode(size)
+		}
+	}
+	if offset < 0 {
+		return nil, errors.New("invalid offset")
+	}
+	if limit < 0 {
+		limit = size - offset
+	}
+	if r, err = o.Object.Open(ctx, options...); err != nil {
+		return nil, err
+	}
+	if offset != 0 || limit < size {
+		// It's a partial read
+		return r, err
+	}
+	return o.f.newHashingReader(ctx, r, func(sums hashMap) {
+		if err := o.putHashes(ctx, sums); err != nil {
+			fs.Infof(o, "auto hashing error: %v", err)
+		}
+	})
+}
+
+type hashingReader struct {
+	rd     io.Reader
+	hasher *hash.MultiHasher
+	fun    func(hashMap)
+}
+
+func (f *Fs) newHashingReader(ctx context.Context, rd io.Reader, fun func(hashMap)) (*hashingReader, error) {
+	hasher, err := hash.NewMultiHasherTypes(f.keepHashes)
+	if err != nil {
+		return nil, err
+	}
+	hr := &hashingReader{
+		rd:     rd,
+		hasher: hasher,
+		fun:    fun,
+	}
+	return hr, nil
+}
+
+func (r *hashingReader) Read(p []byte) (n int, err error) {
+	n, err = r.rd.Read(p)
+	if err != nil && err != io.EOF {
+		r.hasher = nil
+	}
+	if r.hasher != nil {
+		if _, errHash := r.hasher.Write(p[:n]); errHash != nil {
+			r.hasher = nil
+			err = errHash
+		}
+	}
+	if err == io.EOF && r.hasher != nil {
+		r.fun(r.hasher.Sums())
+		r.hasher = nil
+	}
+	return
+}
+
+func (r *hashingReader) Close() error {
+	if rc, ok := r.rd.(io.ReadCloser); ok {
+		return rc.Close()
+	}
+	return nil
+}
+
+// Return object fingerprint or empty string in case of errors
+func (o *Object) fingerprint(ctx context.Context) string {
+	size := o.Object.Size()
```

Comment:
Just wondering why you aren't using `fs.Fingerprint` with `fast=true`?

Reply:
I needed an ad-hoc fingerprint because instead of `baseFs.Hashes().GetOne()` I want `baseFs.HashesNotInHasherCache().GetOne()`

Reply:
I needed an ad-hoc fingerprint because instead of `baseFs.Hashes().GetOne()` I want `baseFs.HashesNotInHasherCache().GetOne()` (and of course none if base hashes are slow).
fs.Fingerprint can't do that.